### ISSUE TITLE
[BUG] TensorDict assignment by index not working properly

### ISSUE THREAD

**Issue by ezhang7423** [facebookresearch/rl#368]
## Describe the bug

When indexing into a tensordict by index and then key, assignment does not work properly. Assignment does appear to work when indexing by key and then index.

## To Reproduce

```

In [1]: from torchrl.data import TensorDict

In [2]: import torch

In [3]: a = TensorDict({'0': torch.zeros((3, 3, 4)), '1': torch.zeros(3, 3)}, batch_size=[3
   ...: , 3])

In [4]: a[0]['0']
Out[4]: 
tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])

In [5]: a['0'][0]
Out[5]: 
tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])

In [6]: a[0]['0'] = torch.full((3, 4), 1)

In [7]: a[0]['0']
Out[7]: 
tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])

In [8]: a['0'][0]
Out[8]: 
tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])

In [9]: a['0'][0] = torch.full((3, 4), 1)

In [10]: a[0]['0']
Out[10]: 
tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]])


In [10]: a['0'][0]
Out[10]: 
tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]])
```

## Expected behavior

I expect assignment to work the same both ways.

## System info

Describe the characteristic of your environment:
 * Describe how the library was installed (pip, source, ...)
 * Python version
 * Versions of any other relevant libraries

```python
import torchrl, numpy, sys
print(torchrl.__version__, numpy.__version__, sys.version, sys.platform)

0.0.1c0 1.23.1 3.8.5 (default, Sep  4 2020, 07:30:14) 
[GCC 7.3.0] linux

```

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
- [x] I have read the [documentation](https://github.com/facebookresearch/rl/tree/main/docs/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

**Comment by vmoens** [comment]
Thanks for reporting this @ezhang7423 
I'm not sure that this is a bug:
```python
a[0]['0'] = torch.full((3, 4), 1)
```
Here you create a new tensordict with data that share the storage with `a`:
```
>>> a[0]["0"]._base is a["0"]
True
```
But when doing
```python
tensordict[key] = something
```
you are basically assigning a new value to a certain key (not writing over the tensor, but assigning a new tensor to a key).
It's kind of a compromise: if we always assume that we are writing things in-place, it will lead to some surprising errors when working with tensors that require gradients etc. (think of it as "what if a[0] was a simple dictionary? Would I expect it to write things in-place on the tensor if I'm not explicitly asking for it?" IMO this should be answered by no but I'm happy to discuss it further :) )

If you want to do this, I can use `SubTensorDict` which will by default write things where needed (or use `set_` and `set_at_`) (this works with #369):

```python
>>> a[0]['0'] = torch.full((3, 4), 1)  # not writing in-place -- expected
>>> a[0]['0']
tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])
>>> a.get_sub_tensordict((0,))["0"] = torch.full((3, 4), 1)  # writing in-place -- expected
>>> a[0]['0']
tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]])
``` 

Hope this is clear.

**Comment by ezhang7423** [comment]
Ahh, that's very interesting! Thank you so much for the detailed response. 

I need to give this a little more thought. Although IMO, I would expect that if a[0] was a simple dictionary, that `a[0]['0']` would do in-place assignment. This is becauase if I thought of `a` just a simple list of dictionaries, then   python  would interpret `a[0]['0'] = 10` as actually changing the value of `a[0]['0']`.  

Could you speak a bit more to the surprising errors that would occur with `requires_grad`?

**Comment by vmoens** [comment]
I will consider this as solved as for PRs #369 but feel free to re-open if you have other comments @ezhang7423
